### ISSUE TITLE
How to recenter?

### ISSUE THREAD

**Issue by friggasaint** [googlevr/cardboard#63]
Has anyone found out how to recenter the camera in VR?
E.g. to a defined orientation?
If not possible yet is there a possibility for this in the future?

**Comment by julien-tamade** [comment]
I found luck by putting the Camera as a child to another gameObject and when I want to recenter the camera, I call

`transform.localRotation = Quaternion.Euler(0.0f, -Camera.main.transform.localRotation.y, 0.0f);`

on the parent gameObject.

Let me know if that helps.

**Comment by jballoffet** [comment]
Sorry for the slow response with this ticket. The Cardboard SDK and the Cardboard XR Plugin for Unity doesn’t have a built in function to recenter the camera. We can add recentering support in the future, this has been added to our backlog. Patches are also welcome.

As @julien-tamade suggested, in the meantime, you can simulate recentering by applying an additional rotation about the Y axis to the head pose to undo the rotation. In Unity, this can be done by making the "head" GameObject (with a TrackedPoseDriver) a child of a "recenter" GameObject and doing "recenterObject.transform.eulerAngles = new Vector3(0, -headObject.transform.eulerAngles.y, 0)".

**Comment by myasovik** [comment]
Hello,

I am using Unity 2019.4 and XR plugin 1.3.0. I am building application for iOS. And the problem I having when you start app and hold your phone vertically in demo scene the cube come up not in front but 90 deg right. If you start the app holding phone horizontally it work good and the cube is in front of the player.

So I want to reset camera Y rotation. But was not able to.

In demo I attached to the "Player" object next script:

```
using UnityEngine;

public class ResetVR : MonoBehaviour
{

    float time;
    void Update()
    {
        if (time < 0)
        {
            transform.localRotation = Quaternion.Euler(0.0f, -Camera.main.transform.localRotation.y, 0.0f);
            time = 5;
        }

        time -= Time.deltaTime;
    }
}
```

It works in UnityEditor but does not have any effect when you actually running app on a phone.

**Comment by Mandelbrow** [comment]
It's pretty useless in Android too. 
I think the recenter method has to reset the localEulerAngles on the camera transform to work as expected.

It can be done by disabling the TrackedPoseDriver, and then resetting the camera-transform before enabling it again. Think it's to expensive to run on each frame, which I has to.

**Comment by myasovik** [comment]
I updated my code, to disable TrackedPoseDriver and move it to MainCamera. It work in UnityEditor but it crash app in iOS. Am I doing something wrong?

```
public class ResetVR : MonoBehaviour
{

    float time;
    void Update()
    {

        if (time < 0)
        {
            GetComponent<UnityEngine.SpatialTracking.TrackedPoseDriver>().enabled = false;
            transform.parent.transform.localRotation = Quaternion.Euler(0.0f, -Camera.main.transform.localRotation.y, 0.0f);
            time = 5;
            GetComponent<UnityEngine.SpatialTracking.TrackedPoseDriver>().enabled = true;
        }

        time -= Time.deltaTime;

    }
}
```

**Comment by Mandelbrow** [comment]
When the tracked pose driver is disabled, the camera local euler angles can be reset to re align it to the parent object it's attached to. Can't see any reason for changing the parents rotation.

Guess it could be implemented as a cheaper native method. Also not sure if there are any extra delay involved when starting the VR tracking agiain.

**Comment by agalbachicar-gg** [comment]
Thanks all for your comments. We are tracking internally this ticket and will prioritise it for the following releases.

**Comment by myasovik** [comment]
I thought that there was a problem in disable TrackedPoseDriver. But I just realized that my demo scene on iOS crashing after reimporting the project.
I did reimport the project, I deleted Xcode project generated by Unity. And the sample scene is crashing all the time at the start. I have version 1.3.0
That is strange as I am positive that it worked before. Did there were pushed some new updates recently?

**Comment by myasovik** [comment]
Hi,
Crash was totally my fault.
Right now I updated to 1.4.0 version.
I try to run code I posted and it working in Unity editor but does not have any effect when I am running on iOS:

`
public class ResetVRCamera : MonoBehaviour
{
    float time;
    void Update()
    {

        if (time < 0)
        {
            GetComponent<UnityEngine.SpatialTracking.TrackedPoseDriver>().enabled = false;
            transform.localRotation = Quaternion.Euler(0.0f, -Camera.main.transform.localRotation.y, 0.0f);
            time = 5;
            GetComponent<UnityEngine.SpatialTracking.TrackedPoseDriver>().enabled = true;
        }

        time -= Time.deltaTime;

    }
}

`

**Comment by Mandelbrow** [comment]
Sorry for the wrong information about just having to .disable, and .enable the TrackedPoseDriver.

It didn't recentered the tracking, as I thought, it just made the camera stop following the head movement, for a while.
The recentering only happened when the XRsubsystem was loaded, i.e. when switching from 3d to vr.

So I guess the feature request remains (haven't tested v1.4.0 though).

**Comment by AbdulQadir91** [comment]
@jballoffet  

> Sorry for the slow response with this ticket. The Cardboard SDK and the Cardboard XR Plugin for Unity doesn’t have a built in function to recenter the camera. We can add recentering support in the future, this has been added to our backlog. Patches are also welcome.
> 
> As @julien-tamade suggested, in the meantime, you can simulate recentering by applying an additional rotation about the Y axis to the head pose to undo the rotation. In Unity, this can be done by making the "head" GameObject (with a TrackedPoseDriver) a child of a "recenter" GameObject and doing "recenterObject.transform.eulerAngles = new Vector3(0, -headObject.transform.eulerAngles.y, 0)".

this is not proper recentering the camera as  gvr sdk do. is there any solution.

**Comment by AwesomelyUp** [comment]
Hello,

Just wanted to check is there any luck or solution for this?

**Comment by myasovik** [comment]
Hello,

Just wanted to check is there any progress or solution for this?

Thank you

**Comment by mfcanovas** [comment]
Just updated my project from an old Unity version with GVR Unity SDK to last Unity version with Cardboard XR Plugin and lost the ability to recenter 😢

XRInputSubsystem.TryRecenter() always returns false

Any workaround for this?

Thanks.

**Comment by jeff-rey-unity** [comment]
You really have to run this in the update function!?

**Comment by jeff-rey-unity** [comment]
For those of you reading this issue, do not put this in the update function.  It works without disabling the TrackedPoseDriver and you call it everytime you need to reset via a button or whatever you want to use.  I am using a button on UI that is in worldspace.   I can get it to reset once wonderfully with the above method that julien-tamade suggested, but the second time (in testing)  it doesn't reset properly.  It's just a math problem I haven't figured out yet, but I will post when I find the solution.

**Comment by jeff-rey-unity** [comment]
I had success with julien-tamade's post the first time, then it did not work the times thereafter.  My current set up is that I have a reset UI button that is gaze based (worldspace) and once the gaze clicks the button(for me it's 3 seconds) I set up a UI that counts down from 3 to give the user time to find the spot they want the UI new position to be.  formally I used the GvrCardboardHelpers.Recenter();. But as you know it does not work with the new set up.  Just like the former posts, Set up a parent object, with the camera as a child.  Notes: you do not have to put the reset in the update function, and you do not need to disable the Tracked pose driver.  just call the code when needed.  `using System.Collections;
using UnityEngine;
using UnityEngine.UI;

public class ResetRotationsssss : MonoBehaviour
{
    //Ui text that displays to the user how much time they have till the reset occurs
    public Text counter;
    // camera's parent
    public Transform camParent;
    //main camera
    public Transform mainCam;
    // main camera's new transform after user looks in the new direction and countdown is zero
    public float camNewCurrent;

    public void ResetRotation()
    {
        StartCoroutine(ResetRot());
    }
   
    public IEnumerator ResetRot()
    {




       

        counter.text = "Ready";
        yield return new WaitForSeconds(1);
        counter.text = "3";
        yield return new WaitForSeconds(1);
        counter.text = "2";
        yield return new WaitForSeconds(1);
        counter.text = "1";
        yield return new WaitForSeconds(1);
        counter.text = "";
        camParent.transform.eulerAngles = new Vector3(0, 0, 0);
        camNewCurrent = mainCam.transform.eulerAngles.y;
        camParent.transform.eulerAngles = new Vector3(0, -camNewCurrent, 0);
        yield return new WaitForEndOfFrame();
        yield return null;

    }
}
`

**Comment by AGamlin** [comment]
I too would like to know what the plan for Recenter() is.

Right now every time my user opens the app, they have to rotate some random rotation away from the position they were when they started the app.

**Comment by agalbachicar** [comment]
We have created a design document proposal: https://docs.google.com/document/d/1ijyL0wXe7XHn1p2YP_bW8Ziaacy7Eto9ZAloPbHokzs/edit?usp=sharing

Please add your thoughts there about the proposed API and functionality. We are also interested in knowing what do you think about the scope (only to yaw) in the recenter feature.

Thanks all for your patience.

**Comment by AGamlin** [comment]
Wow, thanks for going to the effort for getting this drawn up. Honestly,
reading through what you've written, i don't think i'd have anything more
to add. If there is anything else you need in the future please let me
know. Its very important to us that cardboard stays alive and doesnt become
an expensive paperweight like the oculus go and daydream.


Best,

Adam

**Comment by agalbachicar** [comment]
@AGamlin we've learnt with #208  that it is better to go through a design doc first to make sure the implementation is aligned with the feature and API contract first to reduce the burden of developers.

You and all are welcome to send a PR or contribute to that design document with your uncovered use cases. Otherwise, based on prioritisation, this task and others will be implemented by us. At the moment, we are focusing on getting Metal out.

**Comment by jballoffet** [comment]
JIC, the implementation was addressed in #242. We're in the process of merging those changes internally.

**Comment by AGamlin** [comment]
> JIC, the implementation was addressed in #242. We're in the process of merging those changes internally.

thanks, will these changes be available on googlevr/cardboard-xr-plugin for us Unity Devs?

**Comment by xinyunh0929** [comment]
This was fixed in v1.6.0.

**Comment by HgGamer** [comment]
This problem is sill not fixed in 1.7.0

**Comment by rodrigozol** [comment]
I have this problem at last CardBoard SDK! I know how to use the recenter function  Api.Recenter(), in use at CardboardStartup script. My cardboard cam is a child of a car and loses the forward orientation of the father. I am thinking to put recenter triggers at the track as a fake solution. Any real solution?
